M2/Macaulay2/e/interface/cone.cpp

```diff
@@ -0,0 +1,99 @@
+// Written in 2021 by Mahrud Sayrafi
+
+#include "interface/cone.h"
+
+#include <M2/math-include.h>
+
+#include "debug.hpp"
+#include "interface/gmp-util.h"
+#include "interface/matrix.h"
+#include "matrix-con.hpp"
+#include "matrix.hpp"
+#include "relem.hpp"
+
+#include <libnormaliz/cone.h>
+#include <vector>
+
+typedef mpz_class Integer;
+
+/**
+ * \ingroup cones
+ */
+
+const Matrix /* or null */ *rawFourierMotzkin(const Matrix *C)
+{
+  try
+    {
+      // TODO: generalize the input type, in particular to allow lineality space
+      const Ring *R = C->get_ring();
+      const size_t c = C->n_cols();  // rank of ambient lattice
+      const size_t r = C->n_rows();  // number of cone inequalities
+
+      auto ineqs = libnormaliz::Matrix<Integer>(r, c);
```

Comment:
It's slightly better to write ```libnormalize::Matrix<Integer> ineqs(r,c);```. Similarly later for ```libnormalize::Cone<Integer>```. This in theory avoids a copy/move, although here it's almost certainly too cheap to matter.

Reply:
Hmm I think I understand what you mean, but I feel like this is more readable than `ineqs(r, c)` (it almost looks like `ineqs` is a function with two inputs) or `cone(libnormaliz::Type::inequalities, ineqs)`. Do you think the compiler would take care of avoiding this copy/move for us?

Reply:
It's pretty standard c++ notation although for c++11 some suggest to always use ```libnormalize::Matrix<Integer> ineqs{r,c};``` note the braces. This avoids the following pitfall: https://en.wikipedia.org/wiki/Most_vexing_parse.

As for avoiding the copy, usually. In c++17, the compiler will always get rid of the copy (the copy doesn't exist in c++17 because of how the standard implements initialization), but prior to that, it's allowed to and usually it does, but it's not technically required.